pallets/asset/src/lib.rs

```diff
@@ -1586,15 +1746,63 @@ impl<T: Trait> Module<T> {
         Ok(())
     }
 
-    fn _update_checkpoint(ticker: &Ticker, user_did: IdentityId, user_balance: T::Balance) {
+    /// Checks whether a scheduled checkpoint is due for `ticker` and `did`. Returns data read from
+    /// storage that can be reused or `None` if there is no due checkpoint. The function is used to
+    /// check if a protocol fee payment should be made in advance of recording the checkpoint.
+    fn is_checkpoint_due(ticker: &Ticker, did: IdentityId) -> Option<DueCheckpointData> {
+        if Self::is_owner(ticker, did) && NextCheckpoints::contains_key(ticker) {
```

Comment:
I don't think a this function should depend on the `did`. If a checkpoint is "due" we should create a new checkpoint (using the existing functionality), and then every balance will be snapshotted for that checkpoint in `CheckpointBalance`